Issues/862 integrate vue codemirror as editor for object definitions

flx-sta opened the pull request:
fixes #862 
fixes #863  

## Linked PRs
#916 
https://github.com/peregrine-cms/peregrine-cms-e2e-testing/pull/10

## Tasks
- [x] enable file-actions
- [x] make bigger files scrollable
- [x] bind `CommandOrControl + S` to `save()` button function
- [x] auto-format / codeformat (`CommandOrControl` + `Alt`  + `l`) 
- [x] missing folder icon
- [x] jumping always to end of document (no matter where I click)
- [x] ~~moving a "page" node is broken~~ (worked perfectly fine for me)
- [x] missing `publish` action
- [x] rename files name regex seems to be wrong
- [x] ~~(i) button in object-definition files doesn't seem to work properly~~ ([worked fine for me](https://github.com/headwirecom/peregrine-cms/pull/935#issuecomment-885666531))

Commit:
codemirror as editor for object-definitions:

- fix disabled stlying for tab-handlers
- add array for tab-handlers
- add ids object for tabs
- disable all tabs except codemirror for now
- make codemirror default tab

Commit:
Merge branch 'issues/863-allow-file-creation-in-object-definitions' into issues/862-integrate-vue-codemirror-as-editor-for-object-definitions

Commit:
codemirror for objectdefinitionfile

- set codemirror height to 100% in content-view
- remove folder-icon from object-definition files
- add editObjectDefinitionFile stateAction
- add loading file content into form
- add opening in-peregrine editor for objectdefinition file
- todo: save func

Commit:
feature codemirror editing

- fix selecting "code-view" tab bug
- fix styling
- fix objectdefinitionfile being always empty

Commit:
fix: deleting all files instead of the edited

Commit:
remove object definitions nav

reusr1 reviewed (changes requested):
@Felix-Puetz seems there are a couple of issues

- seems you convert the text to json before submitting - I don't think that is necessary as it causes issue when the text in the editor is not a json file
- would be good if the code mirror editor would be the full height
- if I m not mistaken codemirror has code highlighting for json built in

Commit:
make codemirror take up whole screen and add scrollbars

Commit:
object-definition-file

- remove obsolete eslint comment
- json-stringify only if its json

flx-sta commented:
> @Felix-Puetz seems there are a couple of issues
> 
> * seems you convert the text to json before submitting - I don't think that is necessary as it causes issue when the text in the editor is not a json file

* This now only happens if the answer is content-type application/json

> * would be good if the code mirror editor would be the full height

* Applied this as much as its possible (`setSize('100%', '100%')` didn't work.)

> * if I m not mistaken codemirror has code highlighting for json built in

* Unfortunately this is not that easy in peregrine. There is a `/mode` directory in the codemirror folder where you have to require the necessary modes into your file. I've tried exporting these files into the `/etc/felibs/codemiror/mode/` but I wasn't able to load them async. because neither `import()` nor `require()` seem to be supported. I then tried to import it synchronously by using a static `import` at the top of the file but since you have no access to import files from `/node_modules´ this was unsuccessful either

Commit:
file-content always raw

reusr1 commented:
@Felix-Puetz we need to clean up this page a bit
 - remove the tabs
 - add the breadcrumb above code mirror in the header area (makes it easier to navigate)
 - make sure the right pane shows the regular page info relevant to this file type with the regular actions

we also need a better place for the save button. top right is not what we have anywhere else in peregrine.

Commit:
pull/935#issuecomment-855922850

- add missing pathfield
- fix links in pathfield (different behaviour for the file-name itself)
- remove tabs
- remove save button (for now)
- add  stateaction "showFileInfo"
- add fileview as right-panel
- add population file info into right panel

flx-sta commented:
since im implementing a `nt:file` solution for the righ-panel 
- this should also be applied to the secondary files:
  ![image](https://user-images.githubusercontent.com/50131232/121205535-09254400-c878-11eb-8326-fc88f84e6d43.png)

- also these files should open in the form-view (codemirror) too

Commit:
PR #935

- optimize file-editor to actually edit files in general
- fix pathfield
- fix file-preview
- minor fixes

Commit:
add save button (non functional yet

Commit:
pr #935

- generalised fileeditor event further
- add getTenant Mixin
- minimize selectPathInNav code

Commit:
tried to make pageview more dynamic:

make it flexible to change according to its given resourceType (per:page, nt:file) . But it always stops working after the 1. change which seems to do with broken reactivity of the state).
still investigating.

Commit:
feature/codemirror

- fix making pages look selected
- fix reactivity issue for dynamic explorerpreview
- add allowmove condition
- add allowrename condition
- add allowebpublish condition
- add allowdelete condition
- add allowcopy condition
- fix "editEntitiy"
- add "saveFile" state action
- add "editFile" state action
- fix saving files that are not objectdefinitions
- remove "saveObjectDefinitionFile" state action
- remove "editObjectDefinitionFile" state action
- fix z-index of file-save button

Commit:
feature/codemirror

- add async code-mode loading
- add code highlightning
- fix saving files without having to delete them first

reusr1 committed:
added missing parent .content.xml file

Commit:
feature/codemirror

- add save & close button
- add close (without save) button
- fix some styling

Commit:
add '.file-editor' class to file-editor

Commit:
move loading-spinner for file-editor to bottom

Commit:
add code-highlight to objectdefinitions editor

Commit:
remove empty line

Commit:
add css styntax highlighting

Commit:
add jsconfig.json

Commit:
make nodemon more "powerfull"

Commit:
improve nodemon (feedback)

Commit:
fix: make bigger files scrollable #935

Commit:
notifications !

Commit:
add save shortcut

Commit:
fix preventing all hotkeys in codemirror

Commit:
reduce nodemon delay

Commit:
fix editReference

Commit:
add autoformatting #935

Commit:
add file-editor menu (with example actions)

flx-sta commented:
Current state:
![image](https://user-images.githubusercontent.com/50131232/125082131-20e62700-e0c7-11eb-8449-6abf3a0a0c42.png)

Commit:
implement rename file

Commit:
add autofocus for rename modal

Commit:
add class for rename-modal

Commit:
add delete file

Commit:
implement file-copy

Commit:
improve copy-naming (no more overwriting)

Commit:
copy-file state-action improvemenets

Commit:
implement move-file

Commit:
fix "delete-file" in file-editor

Commit:
make menu fix

flx-sta commented:
@reusr1 ready for review

Commit:
fix missing open-folder icon

Commit:
fix: always jumping to the end of the document

Commit:
remove debugging code

Commit:
remove folder_open from nt:file

Commit:
add publish tab to file-explorer

Commit:
add distinguishable class to web-publishing modal

Commit:
Merge branch 'develop-sling12' into issues/862-integrate-vue-codemirror-as-editor-for-object-definitions

Commit:
add missing transition for save&exit icon

golinski commented on admin-base/ui.apps/src/main/content/jcr_root/apps/admin/components/explorer/template.vue:
```diff
@@ -525,11 +528,11 @@ export default {
 
             composumEditable: function(child) {
-                return ['nt:file'].indexOf(child.resourceType) >= 0
+                return [].indexOf(child.resourceType) >= 0
```
`false`?

golinski commented on admin-base/ui.apps/src/main/content/jcr_root/apps/admin/components/explorerpreviewcontent/template.vue:
```diff
@@ -630,12 +647,16 @@ export default {
+        return this.getGeneratedFileSchema();
       } else {
-        return {};
+        if (this.activeTab === Tab.INFO) {
```
`else if` on a single line

golinski commented on admin-base/ui.apps/src/main/content/jcr_root/apps/admin/components/explorerpreviewcontent/template.vue:
```diff
@@ -815,10 +841,25 @@ export default {
+
+        $perAdminApp.stateAction('copyFile', {
+          from: this.currentObject, 
+          to
```
Is this ok? Looks like it should read `to: to`

flx-sta replied:
shorthand for `to: to`

golinski commented on admin-base/ui.apps/src/main/content/jcr_root/apps/admin/components/pageview/template.vue:
```diff
@@ -1,60 +1,80 @@
-      :browserRoot="`${getBasePath()}/pages`"
-      :currentPath="`${getBasePath()}/pages`"
-      :onDelete="onDelete">
+    :model="model"
```
`:key` is missing. Is that expected?

flx-sta replied:
this was resolved in a different way :)

golinski commented on admin-base/ui.apps/src/main/content/jcr_root/apps/admin/components/toolingpage/template.vue:
```diff
@@ -113,25 +113,33 @@ export default {
+            $perAdminApp.stateAction('editFile', {path, resourceType});
           } else {
-            $perAdminApp.stateAction('editPage', target)
+            if (path.startsWith(`/content/${tenant.name}/objects`)) {
```
`else if` on a single line. Couldn't we use the same logic as the first if (instead of checking `path` look at `resourceType`?

flx-sta replied:
that would require alot more refactoring. this is basically "legacy support"

golinski commented on admin-base/ui.apps/src/main/content/jcr_root/content/admin/pages/file/.content.xml:
```diff
@@ -0,0 +1,226 @@
+    <jsonview jcr:primaryType="nt:unstructured"
+      sling:resourceType="admin/components/debugger" dataFrom="/"/>
+
+    <!-- <tour jcr:primaryType="nt:unstructured"
```
Does this commented out fragment serve some purpose?

flx-sta replied:
idk. its not from me and I don't wanna touch it as I try to focus on the changes rather then cleanups in my ticket

golinski commented on admin-base/ui.apps/src/main/js/apiImpl.js:
```diff
@@ -861,7 +861,7 @@ class PerAdminImpl {
   }
 
-  copyPage(srcPath, targetPath) {
+  copyPage(srcPath, targetPath, name=null) {
```
Spaces around `=`? I don't know what prettier likes.

golinski commented on admin-base/ui.apps/src/main/js/perAdminApp.js:
```diff
@@ -301,97 +301,111 @@ function processLoaders(loaders) {
+      }
+    })
+    .then(function() {
+      return api
```
Is the value returned from this anonymous function used? Or is this a prettier idea?

golinski commented on admin-base/ui.apps/src/main/js/perAdminApp.js:
```diff
@@ -425,17 +439,22 @@ function findActionInTree(component, command, target) {
     } else {
-        if(component.$parent === component.$root) {
-            if(!findActionInTree(component.$root, command, target)) {
-                logger.error('action', command, 'not found, ignored, traget was', target)
```
I know this is probably just reformatted, but:
* target instead of traget
* could we use templated strings?

flx-sta replied:
typoo fixed. there is advantages of using `...args` syntax here as the object is not transfered to a string in console but rather displayed indepth

Commit:
remove composum editable

Commit:
fix getSchemaByActiveTab

Commit:
fix editEntity()

Commit:
default value (spaces)

Commit:
fix typo

Commit:
adjust rename validatio for files

flx-sta commented:
@reusr1 (i) icon works fine in `object-definitions` for me

![info-icon-object-definitions](https://user-images.githubusercontent.com/50131232/126794351-c2585caf-921b-47a6-b8bd-cd2563fe16ad.gif)

flx-sta commented:
![image](https://user-images.githubusercontent.com/50131232/126799913-0c1fc238-cc96-4a3d-96d0-c2f1ae99da4f.png)

golinski committed:
use json mode for json files

flx-sta reviewed (commented):
lgtm
